[native] PrestoCpp build from source pipeline

Mionsz opened the pull request:
Fully automated build from source process proposal for presto-native-execution (PrestoCpp and Velox). 
README file added for clarification. appreciate any and all of the feedback.

# Prestissimo - Dockerfile build

> 💡 _**PrestoDB** repository: [Presto - https://github.com/prestodb/presto](https://github.com/prestodb/presto)_

> 💡 _**Velox** repository: [Velox - https://github.com/facebookincubator/velox](https://github.com/facebookincubator/velox)_

## Practical Velox implementation using PrestoCpp

> 📝 _**Note:** This readme and the build process was adapted from internal pipeline. You can e-mail the author if you've got questions [milosz.linkiewicz@intel.com](mailto:milosz.linkiewicz@intel.com)_

Prestissimo, marked in PrestoDB GitHub repository as 'presto-native-execution', is effort of making PrestoDB even better using Velox library as a starting point. Both of mentioned - PrestoCpp and Velox - are mainly written using low level `C` and `C++ 17` languages, which makes the build-from-scratch process humongously complicated. To make this process simple, Intel Cloud Native Data Services Team is introducing 3-stage, fully automated Docker build process based on unmodified project GitHub repository.

## Quick Start

### 1. Clone this repository

```bash
git clone https://github.com/prestodb/presto prestodb
```

### 2. (Optional) Define and export Docker registry, image name and image tag variables

> 📝 _**Note:** Remember to end your `IMAGE_REGISTRY` with `/` as this is required for full tag generation._

> 💡 _**Tip:** Depending on your configuration you may need to run all bellow commands as root user, to switch type as your first command `sudo su`_

> 💡 _**Tip:** If `IMAGE_REGISTRY` is not specified `IMAGE_PUSH` should be set `'0'` or docker image push stage will fail._

Type in you console, changing variables values to meet your needs:

```bash
# defaults to 'avx', more info on Velox GitHub
export CPU_TARGET="avx"
# defaults to 'presto/prestissimo-${CPU_TARGET}-centos'
export IMAGE_NAME='presto/prestissimo-${CPU_TARGET}-centos'
# defaults to 'latest'
export IMAGE_TAG='latest'
# defaults to ''
export IMAGE_REGISTRY='https://my_docker_registry.com/'
# defaults to '0'
export IMAGE_PUSH='0'
```

### 3. Make sure Docker daemon is running

(Ubuntu users) Type in your console:

```bash
systemctl status docker
```

### 4. Build Dockerfile repo

Type in your console:

```bash
cd prestodb/presto-native-execution
make runtime-container
```

The process is fully automated and require no interaction for user. The process of building images for the first time can take up to couple of hours (~1-2h using 10 processor cores).

### 5. Run container

> 📝 _**Note:** Remember that you should start Presto JAVA server first_

Depending on values you have set the container tag is defined as

`PRESTO_CPP_TAG="${IMAGE_REGISTRY}${IMAGE_NAME}:${IMAGE_TAG}"`

for default values this will be just:

`PRESTO_CPP_TAG=presto/prestissimo-avx-centos:latest`

to run container build with default tag execute:

```bash
docker run "presto/prestissimo-avx-centos:latest" \
            --use-env-params \
            --discovery-uri=http://localhost:8080 \
            --http-server-port=8080"
```

to run container interactively, not executing entrypoint file:

```bash
docker run -it --entrypoint=/bin/bash "presto/prestissimo-avx-centos:latest"
```

## Container manual build

For manual build outside Intel network or without access to Cloud Native Data Services Poland Docker registry follow the steps bellow.
In you terminal - with the same session that you want to build the images - define and export environment variables:

```bash
export CPU_TARGET="avx"
export IMAGE_NAME='presto/prestissimo-${CPU_TARGET}-centos'
export IMAGE_TAG='latest'
export IMAGE_REGISTRY='some-registry.my-domain.com/'
export IMAGE_PUSH='0'
export PRESTODB_REPOSITORY=$(git config --get remote.origin.url)
export PRESTODB_CHECKOUT=$(git show -s --format="%H" HEAD)
```

Where `IMAGE_NAME` and `IMAGE_TAG` will be the prestissimo release image name and tag, `IMAGE_REGISTRY` will be the registry that the image will be tagged with and witch will be used to download the images from previous stages in case there are no cached images locally. The `CPU_TARGET` will be unchanged for most of the cases, for more info read the Velox documentation. The `PRESTODB_REPOSITORY` and `PRESTODB_CHECKOUT` will be used as a build repository and branch inside the container. You can set them manually or as provided using git commands.

Then for example to build containers when being behind a proxy server, change dir to and type:

```bash
cd presto-native-execution/scripts/release-centos-dockerfile
docker build \
    --network=host \
    --build-arg http_proxy  \
    --build-arg https_proxy \
    --build-arg no_proxy    \
    --build-arg CPU_TARGET  \
    --build-arg PRESTODB_REPOSITORY \
    --build-arg PRESTODB_CHECKOUT \
    --tag "${IMAGE_REGISTRY}${IMAGE_NAME}:${IMAGE_TAG}" .
```


## Build process - more info - `prestissimo` (with artifacts ~35 GB, without ~10 GB)

Most of runtime and build time dependencies are downloaded, configured and installed in this step. The result from this step is a starting point for both second and third stage. This container will be build 'once per breaking change' in any of repositories. It can be used as starting point for Ci/Cd integrated systems.
This step install Maven, Java 8, Python3-Dev, libboost-dev and lots of other massive frameworks, libraries and applications and ensures that all of steps from 2 stage will run with no errors.

On-top of container from step 1 repository is initialized, Velox and submodules are updated, adapters, connectors and side-dependencies are build and configured. PrestoDB native, full repository build, using Meta wrapper mvnw for Maven is being done. After all of those partial steps, make and build are being run for PrestoCpp and Velox with Parquet, ORC, Hive connector with Thrift with S3-EMRFS filesystem implementation (schema `s3://`) and Hadoop filesystem implementation.

```bash
### DIRECTORY AND MAIN BUILD ARTIFACTS
## Native Presto JAVA build artifacts:
/root/.m2/

## Build, third party dependencies, mostly for adapters
/opt/dependency/
/opt/dependency/aws-sdk-cpp
/opt/dependency/install/
/opt/dependency/install/run/
/opt/dependency/install/bin/
/opt/dependency/install/lib64/

## Root PrestoDB application directory
/opt/presto/

## Root GitHub clone of PrestoDB repository
/opt/presto/_repo/

## Root PrestoCpp subdirectory
/opt/presto/_repo/presto-native-execution/

## Root Velox GitHub repository directory, as PrestoDB submodule
/opt/presto/_repo/presto-native-execution/Velox

## Root build results directory for PrestoCpp with Velox
/opt/presto/_repo/presto-native-execution/_build/release/
/opt/presto/_repo/presto-native-execution/_build/release/velox/
/opt/presto/_repo/presto-native-execution/_build/release/presto_cpp/
```

Release container build - mostly with only the must-have runtime files, including presto_server build presto executable and some libraries. What will be used in the final released container depends on user needs and can be adjusted.

# Prestissimo - runtime configuration and settings

> ⚠️ _**Notice:** Presto-native-execution binary requires 32Gb of RAM at runtime to start (default settings). To override this and overcome runtime error add `node.memory_gb=8` line in `node.properties`.

Presto server with all dependencies can be found inside `/opt/presto/`, runtime name is `presto_server`. There are 2 ways of starting PrestoCpp using provided entry point `/opt/entrypoint.sh`.

## 1) Quick start - pass parameters to entrypoint

This is valid when running using docker and using kubernetes. It is not advised to use this method. User should prefer mounting configuration files using Kubernetes.

```
"/opt/entrypoint.sh --use-env-params --discovery-uri=http://presto-coordinaator.default.svc.cluster.local:8080 --http-server-port=8080"
```


## 2) Using in Kubernetes environment:

Mount config file inside a container as `/opt/presto/node.properties.template`. Replace each variable with you configuration values or leave it as is:

> _**Notice:** set up same values for JAVA coordinator as for prestoCpp - version, location and environment should be the same or you will get connection errors._

```ini
presto.version=0.273.3
node.location=datacenter-warsaw
node.environment=test-environment
node.data-dir=/var/presto/data
catalog.config-dir=/opt/presto/catalog
plugin.dir=/opt/presto/plugin
# node.id is generated and filled during machine startup if not specified
```

Mount config file inside a container as `/opt/presto/config.properties.template`.  Replace each variable with you configuration values:

```ini
coordinator=false
http-server.http.port=8080
discovery.uri=http://presto-coordinaator.default.svc.cluster.local:8080
```

## 3) Hive-Metastore connector and S3 configuration:

For minimum required configuration just mount file `/opt/presto/catalog/hive.properties` inside container at give path (fill `hive.metastore.uri` with you metastore endpoint address):

```ini
connector.name=hive-hadoop2
hive.metastore.uri=thrift://hive-metastore-service.default.svc:9098
hive.pushdown-filter-enabled=true
cache.enabled=true
```

Setting required by S3 connector and Velox query engine, replace with your values, reefer to presto hive connector settings help:
```ini
hive.s3.path-style-access={{ isPathstyle }}
hive.s3.endpoint={{ scheme }}://{{ serviceFqdnTpl . }}:{{ portRest }}
hive.s3.aws-access-key={{ accessKey }}
hive.s3.aws-secret-key={{ secretKey }}
hive.s3.ssl.enabled={{ sslEnabled }}
hive.s3select-pushdown.enabled={{ s3selectPushdownFilterEnabled }}
hive.parquet.pushdown-filter-enabled={{ parquetPushdownFilterEnabled }}
```

![image](https://user-images.githubusercontent.com/89970744/200055119-eb9d67a4-1863-4b27-910a-93b471deb0bd.png)

Signed-off-by: Linkiewicz, Milosz <milosz.linkiewicz@intel.com>

Mionsz commented:
@majetideepak what do you think

Mionsz commented:
@mbasmanova

mbasmanova commented:
CC: @mshang816 @majetideepak Michael, Deepak, would you take a look?

mbasmanova commented:
CC: @raulcd

mbasmanova commented:
CC: @kgpai

Mionsz commented:
This pipeline builds whole PrestoCpp from source in a container. I almost finished modifying the branch from which it is being build now to be included in this PR. I will finish the Job in 2-3 days from now as there is a freed day tomorrow in Poland :)

Mionsz commented:
@mbasmanova I have got first reply from user that managed to build prestoCpp from source using this script - https://github.com/prestodb/presto/issues/18497

mbasmanova commented:
> @mbasmanova I have got first reply from user that managed to build prestoCpp from source using this script - #18497

Nice. What did they say?

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/README.md:
```diff
@@ -0,0 +1,387 @@
+3. Make sure Docker daemon is running in background
+4. Build repo
+
+The process is fully automated and require no interaction for user. The process of building images for the first time can take up to couple of hours (~1-2h using 10 processor cores). `images/prestissimo/build.sh`
```
is this folder correct? `imaged/prestissimo/build.sh`? I can't seem to find it on the repo nor the PR.

Mionsz replied:
Thanks for your review! Fixed

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/README.md:
````diff
@@ -0,0 +1,387 @@
+Then, for example to build final stage container, when being behind a proxy server, type:
+
+```bash
+sudo docker build \
````
`sudo` shouldn't be required unless the user isn't part of the docker group.
Where is the docker build supposed to be done? There are a couple of Dockerfiles, based on the `build-images.sh` I suppose is the `dependencies-dockerfile` but we probably should add that on the README.

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/README.md:
```diff
@@ -0,0 +1,387 @@
+    --build-arg https_proxy \
+    --build-arg no_proxy \
+    --build-arg CPU_TARGET \
+    --build-arg IMAGE_REGISTRY -v .
```
is the `-v` flag required, it fails locally for me with:
```
$ docker --version
Docker version 20.10.21, build baeda1f
```

Mionsz replied:
fixed

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/README.md:
````diff
@@ -0,0 +1,387 @@
+    --build-arg IMAGE_REGISTRY -v .
+```
+
+When building stage 1 remember to tag the image using name `${IMAGE_NAME}-base:${IMAGE_TAG}` so for above environment variables the parameters would look `-t "${IMAGE_REGISTRY}${IMAGE_NAME}-base:${IMAGE_TAG}"`
````
isn't the above command wrong then as it shows:
`-t "${IMAGE_REGISTRY}${IMAGE_NAME}${IMAGE_TAG}"`

raulcd reviewed (commented):
Hi @Mionsz ! Thanks! I tried to build using the README but should we refer on the README to build using the `build-images.sh` script or using the `Makefile` target directly instead of the docker command manually? I tried to follow the README to build but for example there's no reference on how to build the second image (`release-dockerfile/Dockerfile`). I'll give a try to build using the provided Makefile target instead of following the `README`.

Mionsz commented:
> > @mbasmanova I have got first reply from user that managed to build prestoCpp from source using this script - #18497
> 
> Nice. What did they say?

They have managed to build the prestocpp using make command :)

raulcd commented:
> @raulcd I have dramatically simplified the process - can you take a look?

I have built successfully using `make runtime-container`

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/opt/entrypoint.sh:
```diff
@@ -0,0 +1,66 @@
+set -eExv -o functrace
+
+SCRIPT_DIR=$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")
+PRESTO_HOME="${PRESTO_HOME:-"/opt/presto/"}"
```
You are adding the slash everywhere. I think is better to remove it in order to avoid things like:
```
$ cat /opt/presto//config.properties.template
```

```suggestion
PRESTO_HOME="${PRESTO_HOME:-"/opt/presto"}"
```

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/opt/entrypoint.sh:
```diff
@@ -0,0 +1,66 @@
+
+function node_configuration()
+{
+  cat "${PRESTO_HOME}/config.properties.template" > "${PRESTO_HOME}/config.properties"
```
is this expected to be mounted as a volume when running the docker container? Should we copy the default ones on the Dockerfile?
If I try to do a docker run it fails to find it:
```
$ docker run presto/prestissimo-avx-centos:latest
...
cat /opt/presto//config.properties.template
cat: /opt/presto//config.properties.template: No such file or directory
```
If I override the entrypoint to see if the file is there is missing:
```
$ docker run --entrypoint /bin/bash -it presto/prestissimo-avx-centos:latest
bash-4.4$ ls /opt/presto/
_repo  catalog	common	config.properties  connectors  etc  http  libpresto_server_lib.a  operators  presto_server  tests  thrift  types
bash-4.4$ cat /opt/presto/config.properties 
bash-4.4$ 
```
Should we add something on the README on how to run it after built using docker?

Mionsz replied:
node_configuration is designed for use with kubernetes and mounting configuration files as templates that are latet in runtime filled with node-specific values and save as desired configuration files.

Mionsz replied:
To run it using docker one should use something like:
`docker run presto/prestissimo-avx-centos:latest --use-env-params --discovery-uri=http://presto-coordinaator.default.svc.cluster.local:8080 --http-server-port=8080`

raulcd replied:
Hi @Mionsz the work on this PR is great!
Maybe as a new user I would like to have a really simple guide with the two default commands to run locally with docker:
- build --> this is covered with the `make runtime-container` command
- run --> I would like to have a default command saying if you run `docker run xxxxx` this will start a local instance with default settings exposing at port 8080 after building with `make runtime-container`

raulcd replied:
I just saw your previous comment for:
`docker run presto/prestissimo-avx-centos:latest --use-env-params --discovery-uri=http://presto-coordinaator.default.svc.cluster.local:8080 --http-server-port=8080`
this works even though I am getting some failures, I'll investigate:
```
"${PRESTO_HOME}/presto_server" --logtostderr=1 --v=1 "${presto_args[@]}"
+ /opt/presto//presto_server --logtostderr=1 --v=1
I1108 12:57:27.067461    12 PrestoServer.cpp:332] Starting with node memory 40GB
E1108 12:57:27.068285    12 MmapAllocator.cpp:361] mmap failed with 12
E1108 12:57:27.068301    12 Exceptions.h:68] Line: ../../velox/velox/common/memory/MmapAllocator.cpp:362, Function:SizeClass, Expression:  Could not allocate working memorymmap failed with 12, Source: RUNTIME, ErrorCode: INVALID_STATE
E1108 12:57:27.068984    12 ExceptionTracer.cpp:218] terminate() called, exception stack follows
E1108 12:57:27.068993    12 ExceptionTracer.cpp:220] Exception type: facebook::velox::VeloxRuntimeError (11 frames)
    @ 0000000000b47eb5 _ZZN12_GLOBAL__N_111InitializerC4EvENKUlPvPSt9type_infoPPFvS1_EE_clES1_S3_S6_
                       /opt/dependency/folly/_build/../folly/experimental/exception_tracer/ExceptionStackTraceLib.cpp:79
    @ 0000000000b47ee5 _ZZN12_GLOBAL__N_111InitializerC4EvENUlPvPSt9type_infoPPFvS1_EE_4_FUNES1_S3_S6_
                       /opt/dependency/folly/_build/../folly/experimental/exception_tracer/ExceptionStackTraceLib.cpp:79
    @ 0000000000b4ee23 __cxa_throw
                       /opt/dependency/folly/_build/../folly/experimental/exception_tracer/ExceptionTracerLib.cpp:58
    @ 0000000003487376 _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
    @ 000000000347c753 _ZN8facebook5velox6memory13MmapAllocator9SizeClassC2Emm
    @ 000000000347f80b _ZN8facebook5velox6memory13MmapAllocatorC1ERKNS1_20MmapAllocatorOptionsE
    @ 000000000067cd24 _ZN8facebook6presto12PrestoServer20initializeAsyncCacheEv
    @ 000000000067d668 _ZN8facebook6presto12PrestoServer3runEv
    @ 0000000000647104 main
    @ 000000000003ad85 __libc_start_main
    @ 00000000006513ee _start

E1108 12:57:27.103837    12 ExceptionTracer.cpp:222] exception stack complete
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Could not allocate working memorymmap failed with 12
Retriable: False
Function: SizeClass
File: ../../velox/velox/common/memory/MmapAllocator.cpp
Line: 362
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxExceptionC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 3  _ZN8facebook5velox6memory13MmapAllocator9SizeClassC2Emm
# 4  _ZN8facebook5velox6memory13MmapAllocatorC1ERKNS1_20MmapAllocatorOptionsE
# 5  _ZN8facebook6presto12PrestoServer20initializeAsyncCacheEv
# 6  _ZN8facebook6presto12PrestoServer3runEv
# 7  main
# 8  __libc_start_main
# 9  _start

*** Aborted at 1667908647 (Unix time, try 'date -d @1667908647') ***
*** Signal 6 (SIGABRT) (0xba0000000c) received by PID 12 (pthread TID 0x7fc7e3be0e40) (linux TID 12) (maybe from PID 12, UID 186) (code: -6), stack trace: ***
(error retrieving stack trace)
```

Mionsz replied:
Seems like it could not allocate memory, 40GB?

Mionsz replied:
Adding this in config.properties should help
```ini
query.max-memory-per-node=2
query.max-total-memory-per-node=3
```

raulcd replied:
ok, could we add the ability to configure the memory allocated to be overridden too? This worked locally:
```
printf "node.memory_gb=${memory_gb}\n" >> "${PRESTO_HOME}/node.properties"
```

Mionsz replied:
This should be done by config file mounted inside the container. We should not override the defaults using scripting I think.

Mionsz replied:
It would be better to add info that default configuration requires 40GB of memory to be reserved by Prestissimo, as this is a configuration setting.

raulcd commented on presto-native-execution/scripts/release-centos-dockerfile/opt/entrypoint.sh:
```diff
@@ -0,0 +1,66 @@
+
+function node_configuration()
+{
+  cat "${PRESTO_HOME}/config.properties.template" > "${PRESTO_HOME}/config.properties"
+  cat "${PRESTO_HOME}/node.properties.template" > "${PRESTO_HOME}/node.properties"
```
ok, I think it was failing to find because this is copied on the Dockerfile under `/opt/presto/etc` not under `/opt/presto`, right?
```suggestion
  cat "${PRESTO_HOME}/etc/config.properties.template" > "${PRESTO_HOME}/config.properties"
  cat "${PRESTO_HOME}/etc/node.properties.template" > "${PRESTO_HOME}/node.properties"
```

Mionsz replied:
I have fixed this to run with defaults if no config is attached. The above suggestion is not valid to this workflow - where use should mount config templates to presto home dir.

Mionsz replied:
https://github.com/prestodb/presto/pull/18572/commits/23c0784c17115047f0d67af901f5360ee1190171

Mionsz commented on presto-native-execution/scripts/release-centos-dockerfile/Dockerfile:
```diff
@@ -0,0 +1,162 @@
+
+USER ${PRESTO_UID}
+
+COPY --chown=${PRESTO_UID} etc/ /opt/presto/etc/
```
No this is not wrong, it should be mounted in Kubernetes like:

```yaml
        volumeMounts:
          - name: presto-config
            mountPath: /opt/presto/config.properties.template
            subPath: worker.config.properties
          - name: presto-config
            mountPath: /opt/presto/node.properties.template
            subPath: node.properties
```

and now if this is not found the etc files will be used.

Mionsz replied:
This is described in Kubernetes environment part of readme. But as I mentioned in the beginning this PR is adopting our internal pipeline build process so maybe we should change so things to meet opensource community requirements :)

kgpai commented on presto-native-execution/scripts/release-centos-dockerfile/Dockerfile:
```diff
@@ -0,0 +1,162 @@
+USER root:root
+
+ENV CPU_TARGET=${CPU_TARGET}
+ENV TZ=Europe/Warsaw
```
can we get this from the env or say use timedatectl ?

Mionsz replied:
IT is no longer required. Removed.

kgpai commented on presto-native-execution/scripts/release-centos-dockerfile/Dockerfile:
```diff
@@ -0,0 +1,162 @@
+#-----------------------------------------------------------
+WORKDIR ${PRESTODB_HOME}
+RUN set -exu && \
+    dnf -y install git wget unzip pkg-config python3-devel java clang-tools-extra bzip2 bzip2-libs && \
```
Convention in Velox so far has been to have fairly light weight dockerfiles , which call into scripts to do the heavy lifting regarding installation of dependencies etc. This is because its easier to validate/remove these dependencies from these script files .

Mionsz replied:
True. Fixed.

Mionsz replied:
I have only left the minimum required in case the source image changes.

kgpai commented on presto-native-execution/scripts/release-centos-dockerfile/README.md:
```diff
@@ -0,0 +1,492 @@
+
+> 📝 _**Note:** This readme and the build process was adapted from internal pipeline. You can e-mail the author if you've got questions [milosz.linkiewicz@intel.com](mailto:milosz.linkiewicz@intel.com)_
+
+Prestissimo, marked in PrestoDB GitHub repository as 'presto-native-execution', is effort of making PrestoDB even better using Velox library as a starting point. Both of mentioned - PrestoCpp and Velox - are mainly written using low level `C` and `C++ 17` languages, which makes the build-from-scratch process humongously complicated. To make this process simple, Intel Cloud Native Data Services Te ...
```
> which makes the build-from-scratch process humongously complicated

We are working hard to simplify this though :)

Mionsz replied:
I know :-p, but most of the folks don't know C++ at all so this addresses them.

kgpai reviewed (commented):
Thanks for this PR! Would it be possible to set this up as a nightly job to validate that changes in velox dont break presto_cpp.

Mionsz commented on presto-native-execution/scripts/build-centos.sh:
```diff
@@ -0,0 +1,112 @@
+    error=0
+    check=$(txt_green success)
+    prompt "Velox build requires bellow CPU instructions to be available:"
+    for flag in 'bmi|bmi1' 'bmi2' 'f16c';
```
Fixed. Yes. I have made this only informational now - so that the user know what can be source of an issue in case the compiler  complains further in the code.

Mionsz commented on presto-native-execution/scripts/build-centos.sh:
```diff
@@ -0,0 +1,112 @@
+set -eE -o pipefail
+trap 'error "Stage failed, exiting"; exit 5' SIGSTOP SIGINT SIGTERM SIGQUIT ERR
+
+print_logo
```
Removed.

Mionsz commented on presto-native-execution/scripts/release-centos-dockerfile/opt/common.sh:
```diff
@@ -0,0 +1,188 @@
+    printf '%s\n' "${_output_array[@]}" >&2
+}
+
+function print_logo() {
```
Removed.

Mionsz commented:
> Thanks for this PR! Would it be possible to set this up as a nightly job to validate that changes in velox dont break presto_cpp.

Of course. But this would be a separate PR.

kgpai reviewed (commented):
I tried building this on my mac; obviously it fails since it relies on /proc/cpuinfo . Can we try making it work on the mac too ?

Mionsz commented on presto-native-execution/scripts/build-centos.sh:
```diff
@@ -0,0 +1,115 @@
+        echo $(cat /proc/cpuinfo) | grep -E -q " $flag " && check=$(txt_green success) || check=$(txt_red failed)
+        prompt "Testing (${flag}): \t$check"
+    done
+    prompt "Velox build suggest bellow CPU instructions to be available:"
```
Sorry for long response but I had a vacation. I have made the tests able to fail now, we can add macOS tests but not sure if it is relevant to this PR.

Mionsz replied:
Ok. I have added a simple MacOS workflow in the bash scripting. Should work, but would need some feedback.

kgpai replied:
Thanks @Mionsz . I will test it on a mac, Let me know if its ok to push fixes to this branch if I find bugs ?

kgpai commented:
@Mionsz @raulcd Testing this locally how do you deal with gits key verification failed errors - that prevent it from cloning repo's ?

kgpai commented:
@Mionsz  Doesnt look like its trying to checkout any special commit. 
```
#9 84.53 + git clone --progress git@github.com:prestodb/presto.git /opt/presto/_repo
#9 84.53 Cloning into '/opt/presto/_repo'...
#9 84.86 Host key verification failed.
#9 84.86 fatal: Could not read from remote repository.
```

I am digging into it more though..

Fiooodooor commented:
> @Mionsz Doesnt look like its trying to checkout any special commit.
> 
> ```
> #9 84.53 + git clone --progress git@github.com:prestodb/presto.git /opt/presto/_repo
> #9 84.53 Cloning into '/opt/presto/_repo'...
> #9 84.86 Host key verification failed.
> #9 84.86 fatal: Could not read from remote repository.
> ```
> 
> I am digging into it more though..

I did not take that in consideration - the possibility of user using SSH based git origin. I will correct this later. This should be added in dockerfile - for example line 41:
```bash
ENV GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no"
```

Mionsz commented:
@mbasmanova @majetideepak @raulcd @kgpai are we considering moving this forward and addressing issues in new PR as this is becoming big and hard to maintain?

Mionsz commented:
@kesavkolla Try with this build and latest commits as those fixes lots of possible issues

kgpai commented:
@Mionsz Thanks, I get another failure now though : 
```
#11 951.2 make[1]: Leaving directory '/opt/dependency/gperf-3.1/doc'
#11 951.2 + ln -s /usr/local/gperf/3_1/bin/gperf /usr/local/bin/
#11 951.2 + git clone https://github.com/facebook/folly
#11 951.2 fatal: destination path 'folly' already exists and is not an empty directory.
```
This happens when building presto_native after velox. Recently some changes readded folly to the setup so I believe thats the reason for this failure. 

> @mbasmanova @majetideepak @raulcd @kgpai are we considering moving this forward and addressing issues in new PR as this is becoming big and hard to maintain?
I think we should attempt to get this PR in a good shape before we can merge. If its

Mionsz commented:
> @Mionsz Thanks, I get another failure now though :
> 
> ```
> #11 951.2 make[1]: Leaving directory '/opt/dependency/gperf-3.1/doc'
> #11 951.2 + ln -s /usr/local/gperf/3_1/bin/gperf /usr/local/bin/
> #11 951.2 + git clone https://github.com/facebook/folly
> #11 951.2 fatal: destination path 'folly' already exists and is not an empty directory.
> ```
> 
> This happens when building presto_native after velox. Recently some changes readded folly to the setup so I believe thats the reason for this failure.
> 
> > @mbasmanova @majetideepak @raulcd @kgpai are we considering moving this forward and addressing issues in new PR as this is becoming big and hard to maintain?
> > I think we should attempt to get this PR in a good shape before we can merge.

Nah, you should fetch the repository and use this commit for building - i have fixed this couple of hours ago.  They have reverted the folly build in scripting - so the `!` notation in our centos build is enough, it will ignore Folly clone errors and continue building.
![image](https://user-images.githubusercontent.com/89970744/202869221-33f9bd78-b56b-45bf-8367-1389dc5ea599.png)

Mionsz commented:
@raulcd Could you try building it now? I have made simple fix.
I had the same issue some time ago and as far as I remember adding `-DBUILD_SHARED_LIBS=ON` should help but that would need a Velox PR. I have just added a sed command for commenting Folly wget and cmak lines from Velox and it should be build by PrestoCpp script with proper flags. 

There is a real need of pipeline between Velox and PrestoCpp so that one does not break another, this starts to get annoying as main focus seems like keeping up with and fixing the breaking changes :-/

kgpai commented:
@Mionsz I was able to build the docker image fine with your latest changes. I am going to test the image out.

Mionsz commented:
@raulcd no problem, I am glad it helped :-)!
So you hade to add something like `node.memory_gb=8` in `node.properties` yes? As the default is set to 32 as far as I remember. I will include it later in README as this can be a common type of error.

raulcd commented:
@Mionsz yes. I added this to `node.properties` via the `entrypoint.sh` file:
```
+++ b/presto-native-execution/scripts/release-centos-dockerfile/opt/entrypoint.sh
@@ -55,6 +55,7 @@ function node_command_line_config()
   printf "node.location=torun-cluster\n"  >> "${PRESTO_HOME}/node.properties"
   printf "node.id=${NODE_UUID}\n"        >> "${PRESTO_HOME}/node.properties"
   printf "node.ip=$(hostname -I)\n"      >> "${PRESTO_HOME}/node.properties"
+  printf "node.memory_gb=8\n" >> "${PRESTO_HOME}/node.properties"
 }
 ```

Mionsz commented:
@raulcd check the common.sh and entrypoint.sh now :-). I have added automated way for handling the memory 32 GB constrain.

Leaving the ENV variable undefined allows script for probing a valid for runtime machine configuration by trying to lower node memory starting from 32 down to 4.
If the value was set in any way - by parameter, env variable (script strictly test is value is valid) or when attaching prefilled template - script ignores and does not override.

kgpai commented:
@Mionsz Was able to test this out and works as expected.

kgpai commented:
@Mionsz Can you fix/validate that the other modules failure is unrelated ?

Mionsz commented:
@kgpai Yes. Let us wait for the build result and In case something failed I will check/fix it.

Mionsz commented:
@kgpai The only one seems to have an issue is a Maven JAVA test. It is totally unrelated as no changes beyond presto-native-execution were made:
![image](https://user-images.githubusercontent.com/89970744/204347553-483a169c-61b2-4c5b-85fc-536c2ebfaa7d.png)

Mionsz commented:
@kgpai Notice there is no longer warning/issue with fbthrift present in other builds:

https://github.com/prestodb/presto/actions/runs/3566915657/jobs/5993964104#step:8:67
```shell
-- Found FBThrift: /usr/local
```

https://github.com/prestodb/presto/actions/runs/3566313044/jobs/5992587501#step:8:66
```shell
CMake Warning at CMakeLists.txt:159 (find_package):
  By not providing "FindFBTHRIFT.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "FBTHRIFT",
  but CMake did not find one.

  Could not find a package configuration file provided by "FBTHRIFT" with any
  of the following names:

    FBTHRIFTConfig.cmake
    fbthrift-config.cmake

  Add the installation prefix of "FBTHRIFT" to CMAKE_PREFIX_PATH or set
  "FBTHRIFT_DIR" to a directory containing one of the above files.  If
  "FBTHRIFT" provides a separate development package or SDK, be sure it has
  been installed.
```

Mionsz commented:
@kgpai all other green :-)
